File: packages/python/anip-server/src/anip_server/manifest.py
"""Manifest builder for ANIP services."""
from __future__ import annotations

import hashlib
import json
from datetime import datetime, timedelta, timezone

from anip_core import (
    ANIPManifest,
    CapabilityDeclaration,
    ManifestMetadata,
    ProfileVersions,
    ServiceIdentity,
    TrustPosture,
    PROTOCOL_VERSION,
    DEFAULT_PROFILE,
)


def build_manifest(
    *,
    capabilities: dict[str, CapabilityDeclaration],
    trust: TrustPosture,
    service_identity: ServiceIdentity,
    expires_days: int = 30,
) -> ANIPManifest:
    """Build an ANIP manifest from parameters (no env-var parsing)."""
    manifest = ANIPManifest(
        protocol=PROTOCOL_VERSION,
        profile=ProfileVersions(**DEFAULT_PROFILE),
        capabilities=capabilities,
        service_identity=service_identity,
        trust=trust,
    )

    caps_json = json.dumps(
        {k: v.model_dump() for k, v in capabilities.items()},
        sort_keys=True,
    ).encode()
    now = datetime.now(timezone.utc)
    manifest.manifest_metadata = ManifestMetadata(
        sha256=hashlib.sha256(caps_json).hexdigest(),
        issued_at=now.isoformat(),
        expires_at=(now + timedelta(days=expires_days)).isoformat(),
    )

    return manifest

File: packages/typescript/server/src/manifest.ts
/**
 * Manifest builder for ANIP services.
 *
 * Fully parameterized — no environment variable parsing.
 */

import { createHash } from "crypto";
import {
  PROTOCOL_VERSION,
  MANIFEST_VERSION,
  DEFAULT_PROFILE,
} from "@anip/core";
import type {
  ANIPManifest as ANIPManifestType,
  CapabilityDeclaration as CapabilityDeclarationType,
  TrustPosture as TrustPostureType,
  ServiceIdentity as ServiceIdentityType,
} from "@anip/core";

export interface BuildManifestOpts {
  capabilities: Record<string, CapabilityDeclarationType>;
  trust: TrustPostureType;
  serviceIdentity: ServiceIdentityType;
  expiresDays?: number;
}

/**
 * Build an ANIP manifest from parameters (no env-var parsing).
 */
export function buildManifest(opts: BuildManifestOpts): ANIPManifestType {
  const { capabilities, trust, serviceIdentity, expiresDays = 30 } = opts;

  const capsJson = JSON.stringify(capabilities, Object.keys(capabilities).sort());
  const sha256 = createHash("sha256").update(capsJson).digest("hex");

  const now = new Date();
  const expiresAt = new Date(
    now.getTime() + expiresDays * 24 * 60 * 60 * 1000,
  );

  return {
    protocol: PROTOCOL_VERSION,
    profile: { ...DEFAULT_PROFILE },
    capabilities,
    manifest_metadata: {
      version: MANIFEST_VERSION,
      sha256,
      issued_at: now.toISOString(),
      expires_at: expiresAt.toISOString(),
    },
    service_identity: serviceIdentity,
    trust,
  } as ANIPManifestType;
}
